File: Find pairs with given sum in Doubly Linked List.java

class Solution {
    public static ArrayList<ArrayList<Integer>> findPairsWithGivenSum(int target, Node head) {
        // code here
        Node tail=head;
        while(tail.next!=null)tail=tail.next;
        ArrayList<ArrayList<Integer>> ans=new ArrayList<>();
        while(head!=tail){
            if(head.data+tail.data==target){
                ArrayList<Integer> pair=new ArrayList<>();
                pair.add(head.data);
                pair.add(tail.data);
                ans.add(pair);
                tail=tail.prev;
            }else if(head.data+tail.data>target){
                tail=tail.prev;
            }else{
                head=head.next;
            }
        }
        return ans;
    }

File: Find pairs with given sum in doubly linked list.cpp
class Solution
{
public:
    vector<pair<int, int>> findPairsWithGivenSum(Node *head, int target)
    {
        // code here
        vector<pair<int,int>> ans;
        
        Node* front = head;
        Node* rear = head;
        
        while(rear->next)
        rear = rear->next;
        
        while(front->data < rear->data)
        {
            int num1 = front->data, num2 = rear->data;
            
            if(num1 + num2 == target)
            {
                ans.push_back({num1, num2});
                front = front->next;
            }
            
            else if(num1 + num2 > target)
                rear = rear->prev;
                
            else
                front = front->next;
        }
        
        return ans;
    }
};
